lib/Driver/ToolChains/Flang.cpp

```diff
@@ -294,28 +294,34 @@ void FlangFrontend::ConstructJob(Compilation &C, const JobAction &JA,
     CommonCmdArgs.push_back("8");
   }
 
-  // -g should produce DWARFv2
-  for (auto Arg : Args.filtered(options::OPT_g_Flag)) {
-    Arg->claim();
-    CommonCmdArgs.push_back("-x");
-    CommonCmdArgs.push_back("120");
-    CommonCmdArgs.push_back("0x200");
-  }
+  // Last argument of -g/-gdwarfX should be taken.
+  Arg *GArg = Args.getLastArg(options::OPT_g_Flag);
+  Arg *GDwarfArg = Args.getLastArg(options::OPT_gdwarf_2,
+                                   options::OPT_gdwarf_3,
+                                   options::OPT_gdwarf_4,
+                                   options::OPT_gdwarf_5);
+
+  if (GArg || GDwarfArg) {
+
+    for (auto Arg : Args.filtered(options::OPT_g_Flag, options::OPT_gdwarf_2,
+                                  options::OPT_gdwarf_3, options::OPT_gdwarf_4,
+                                  options::OPT_gdwarf_5)) {
+      Arg->claim();
+    }
 
-  // -gdwarf-2
-  for (auto Arg : Args.filtered(options::OPT_gdwarf_2)) {
-    Arg->claim();
     CommonCmdArgs.push_back("-x");
     CommonCmdArgs.push_back("120");
-    CommonCmdArgs.push_back("0x200");
-  }
 
-  // -gdwarf-3
-  for (auto Arg : Args.filtered(options::OPT_gdwarf_3)) {
-    Arg->claim();
-    CommonCmdArgs.push_back("-x");
-    CommonCmdArgs.push_back("120");
-    CommonCmdArgs.push_back("0x4000");
+    if (!GDwarfArg) // -g should produce default (DWARFv4)
+      CommonCmdArgs.push_back("0x1000000");
+    else if (GDwarfArg->getOption().matches(options::OPT_gdwarf_2)) // -gdwarf-2
+      CommonCmdArgs.push_back("0x200");
+    else if (GDwarfArg->getOption().matches(options::OPT_gdwarf_3)) // -gdwarf-3
+      CommonCmdArgs.push_back("0x4000");
+    else if (GDwarfArg->getOption().matches(options::OPT_gdwarf_4)) // -gdwarf-4
+      CommonCmdArgs.push_back("0x1000000");
```

Comment:
Is this logic correct? It looks like "no -g" or "-gdwarf 4" will pass `-x 120 0x1000000`, but shouldn't that x-flag also be passed if -g appears last on the command line?

Reply:
Repeating the reply for another comment.


-g suggests dwarf version implicitly, while -gdwarf-N suggests dwarf version explicitly. Implicit suggestion of version is taken only when no explicit suggestion is present. This behavior is exhibited for gfortran and clang as well.
Ex

$ gfortran -gdwarf-2 -g my.f90
$ llvm-dwarfdump a.out | grep version
0x00000000: Compile Unit: length = 0x000000a5 version = 0x0002 abbr_offset = 0x0000 addr_size = 0x08 (next unit at 0x000000a9)
$ gfortran -g my.f90
$ llvm-dwarfdump a.out | grep version
0x00000000: Compile Unit: length = 0x0000009d version = 0x0004 abbr_offset = 0x0000 addr_size = 0x08 (next unit at 0x000000a1)
$ clang -gdwarf-2 -g my.c
$ llvm-dwarfdump a.out | grep version
0x00000000: Compile Unit: length = 0x00000050 version = 0x0002 abbr_offset = 0x0000 addr_size = 0x08 (next unit at 0x00000054)
$ clang -g my.c
$ llvm-dwarfdump a.out | grep version
0x00000000: Compile Unit: length = 0x00000047 version = 0x0004 abbr_offset = 0x0000 addr_size = 0x08 (next unit at 0x0000004b)

Reply:
This is a different question, though.

Shouldn't "flang -g" emit `-x 120 0x1000000`?

Reply:
I am sorry, I did not understand question earlier.
Yes, flang -g should emit -x 120 0x1000000 and it does so.
If the control reaches at line#315 then it must have satisfied the condition at line#304.
line# 304 checks whether any of -g or -gdwarf-N are mentioned.
at line#315 if -gdwarf-N is found not mentioned then -g must have been mentioned (alone).
This condition will satisfy for "flang -g" and "-x 120 0x1000000" will be emitted.

Reply:
Ah. Got it, thanks!

Can you expand the comment on line 315 to read: "-g without -gdwarf-X produces default (DWARFv4)"?

Reply:
Thanks for your comment. I shall do that.